counter/src/main/java/org/infinispan/counter/impl/manager/EmbeddedCounterManager.java

```diff
@@ -87,57 +81,77 @@ private static <T> T validateCounter(Class<T> tClass, Object retVal) {
+
+               if (config.type() == CounterType.WEAK) {
+                  return weakCounterFactory.removeWeakCounter(counterName, config);
+               } else {
+                  return strongCounterFactory.removeStrongCounter(counterName);
+               }
+            });
+      if (keepConfig) {
+         return removeStage;
       }
-      counters.compute(counterName, (name, counter) -> {
-         removeCounter(name, counter, configuration);
-         if (!keepConfig) awaitCounterOperation(configurationManager.removeConfiguration(name));
-         return null;
-      });
+      return removeStage.thenCompose(unused -> configurationManager.removeConfiguration(counterName))
+            .thenApply(toNullFunction());
    }
 
    @Override
    public void undefineCounter(String counterName) {
-      removeCounter(counterName, false);
+      awaitCounterOperation(removeAsync(counterName, false));
    }
 
    @Override
    public StrongCounter getStrongCounter(String name) {
-      checkStarted();
-      Object counter = counters.computeIfAbsent(name, this::createCounter);
-      return validateCounter(StrongCounter.class, counter);
+      return awaitCounterOperation(getStrongCounterAsync(name));
    }
 
-   public StrongCounter getCreatedStrongCounter(String name) {
-      checkStarted();
-      Object counter = counters.get(name);
-      if (counter == null) {
-         return null;
-      }
-      return validateCounter(StrongCounter.class, counter);
+   public CompletionStage<StrongCounter> getStrongCounterAsync(String counterName) {
+      return getOrCreateAsync(counterName).thenApply(InternalCounterAdmin::asStrongCounter);
    }
 
    @Override
    public WeakCounter getWeakCounter(String name) {
-      checkStarted();
-      Object counter = counters.computeIfAbsent(name, this::createCounter);
-      return validateCounter(WeakCounter.class, counter);
+      return awaitCounterOperation(getWeakCounterAsync(name));
    }
 
-   public WeakCounter getCreatedWeakCounter(String name) {
-      checkStarted();
-      Object counter = counters.get(name);
-      if (counter == null) {
-         return null;
+   public CompletionStage<WeakCounter> getWeakCounterAsync(String counterName) {
+      return getOrCreateAsync(counterName).thenApply(InternalCounterAdmin::asWeakCounter);
+   }
+
+   public CompletionStage<InternalCounterAdmin> getOrCreateAsync(String counterName) {
+      if (stopped) {
+         return completedExceptionFuture(CONTAINER.counterManagerNotStarted());
+      }
+      CompletableFuture<InternalCounterAdmin> stage = counters.computeIfAbsent(counterName, this::createCounter);
+
+      if (CompletionStages.isCompletedSuccessfully(stage)) {
+         // avoid invoking "exceptionally()" every time
+         return stage;
       }
-      return validateCounter(WeakCounter.class, counter);
+
+      // remove if it fails
+      stage.exceptionally(throwable -> {
+         counters.remove(counterName, stage);
```

Comment:
Should we log something here?

Reply:
we never did before. TBH, I'm neutral since the invoker will get the exception right away.

Reply:
Alright.